src/main/scala/ohnosequences/awstools/s3/S3.scala

```diff
@@ -1,45 +1,52 @@
 package ohnosequences.awstools.s3
 
-import java.io.{IOException, InputStream, ByteArrayInputStream, File}
-import java.net.URL
 
-import ohnosequences.awstools.regions.Region._
+import ohnosequences.awstools.regions._
+import ohnosequences.logging.Logger
 
 import com.amazonaws.auth._
 import com.amazonaws.services.s3.{AmazonS3, AmazonS3Client}
-import com.amazonaws.services.s3.model._
-import ohnosequences.logging.Logger
-
-import scala.collection.JavaConversions._
+import com.amazonaws.services.s3.model.{ Region => _ , _ }
 import com.amazonaws.services.s3.transfer.{Transfer, TransferManager}
-
 import com.amazonaws.{AmazonClientException, AmazonServiceException}
-import scala.collection.mutable.ListBuffer
 import com.amazonaws.internal.StaticCredentialsProvider
 import com.amazonaws.event._
 import com.amazonaws.event.{ProgressListener => PListener, ProgressEvent => PEvent}
 
+import scala.collection.JavaConversions._
+import scala.collection.mutable.ListBuffer
 import scala.concurrent.duration.Duration
+
 import scala.util.{Failure, Success, Try}
 
+import java.io.{IOException, InputStream, ByteArrayInputStream, File}
+import java.net.URI
+import java.net.URL
+
 
 sealed trait AnyS3Address {
   val bucket: String
   val key: String
 
+  val segments: Seq[String]
+
+  @deprecated("Use toURI method instead, or just toString", since = "v0.17.0")
   final def url = "s3://" + bucket + "/" + key
-  override def toString = url
-}
 
+  def toURI: URI = new URI("s3", bucket, s"/${key}", null)
+
+  override def toString = toURI.normalize.toString
+
+  def toHttpsURL(region: Region): URL = new URL("https", s"s3-${region}.amazonaws.com", s"${bucket}/${key}")
+}
 
 case class S3Folder(b: String, k: String) extends AnyS3Address {
   val bucket = b.stripSuffix("/")
```

Comment:
bucket name restrictions are a bit more complex than this, see http://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html#bucketnamingrules It would be good to have at least an approximation to that.